apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs

```diff
@@ -383,71 +478,125 @@ internal PublisherClientImpl(TopicName topicName, IEnumerable<PublisherServiceAp
         private readonly IScheduler _scheduler;
         private readonly TaskHelper _taskHelper;
         private readonly Func<Task> _shutdown;
+        private readonly bool _enableMessageOrdering;
 
         // Batching settings
         private readonly long _batchElementCountThreshold;
         private readonly long _batchByteCountThreshold;
         private readonly TimeSpan? _batchDelayThreshold;
 
-        //Internal state
+        // Internal state
+        // Clients idle, ready to be used.
         private readonly Queue<PublisherServiceApiClient> _idleClients;
-        private readonly Queue<Batch> _batchesReady;
-        private readonly CancellationTokenSource _softStopCts;
-        private readonly CancellationTokenSource _hardStopCts;
-        private readonly TaskCompletionSource<int> _shutdownTcs;
-        private Batch _currentBatch;
+        // Batches ready to send. Will only ever include at most one batch per ordering-key.
+        private readonly Queue<ReadyBatch> _batchesReady;
+        // First: Currently-filling batch; Last: Next batch to send.
+        private readonly Dictionary<string, KeyState> _keyedState;
```

Comment:
Presumably all state is guarded by the lock, so nothing needs to be thread-safe?